packages/ckeditor5-engine/tests/model/utils/insertcontent.js

```diff
@@ -1299,128 +1302,1629 @@ describe( 'DataController utils', () => {
+					}, [ 1 ] );
+
+					const expectedMarker = model.markers.get( 'marker-a' );
+
+					expect( getData( model ) ).to.equal( '<paragraph>foo[]</paragraph><imageBlock></imageBlock>' );
+
+					expect( expectedMarker.getRange().start.path ).to.deep.equal( [ 1 ] );
+					expect( expectedMarker.getRange().end.path ).to.deep.equal( [ 1 ] );
+				} );
+
+				it( 'should create collapsed marker after inserting imageBlock before paragraph', () => {
+					// []<paragraph>foo</paragraph>
+					// {}<imageBlock></imageBlock>
+					//
+					// {}<imageBlock></imageBlock><paragraph>foo</paragraph>
+					setData( model, '[]<paragraph>foo</paragraph>' );
+
+					insertHelper( '<imageBlock></imageBlock>', {
+						'marker-a': { start: [ 0 ], end: [ 0 ] }
+					}, [ 0 ] );
+
+					const expectedMarker = model.markers.get( 'marker-a' );
+
+					expect( getData( model ) ).to.equal( '<imageBlock></imageBlock><paragraph>[]foo</paragraph>' );
+
+					expect( expectedMarker.getRange().start.path ).to.deep.equal( [ 0 ] );
+					expect( expectedMarker.getRange().end.path ).to.deep.equal( [ 0 ] );
+				} );
+
+				it( 'should create multiple collapsed marker after inserting different elements in empty content', () => {
+					// []
+					// {}<imageInline></imageInline>()<imageBlock></imageBlock>
+					//
+					// <paragraph>{}<imageInline></imageInline>()</paragraph><imageBlock></imageBlock>
+					setData( model, '[]' );
+
+					insertHelper( '<imageInline></imageInline><imageBlock></imageBlock>', {
+						'marker-a': { start: [ 0 ], end: [ 0 ] },
+						'marker-b': { start: [ 1 ], end: [ 1 ] }
+					} );
+
+					const expectedMarkerA = model.markers.get( 'marker-a' );
+					const expectedMarkerB = model.markers.get( 'marker-b' );
+
+					expect( getData( model ) ).to.equal( '<paragraph><imageInline></imageInline></paragraph>[<imageBlock></imageBlock>]' );
+
+					expect( expectedMarkerA.getRange().start.path ).to.deep.equal( [ 0, 0 ] );
+					expect( expectedMarkerA.getRange().end.path ).to.deep.equal( [ 0, 0 ] );
+					expect( expectedMarkerB.getRange().start.path ).to.deep.equal( [ 0, 1 ] );
+					expect( expectedMarkerB.getRange().end.path ).to.deep.equal( [ 0, 1 ] );
+				} );
+			} );
+
+			it( 'should not create marker after inserting empty document fragment', () => {
+				// <paragraph>fo[]o</paragraph>
+				// '{}'
+				//
+				// <paragraph>foo</paragraph>
+				setData( model, '<paragraph>fo[]o</paragraph>' );
+
+				insertHelper( new DocumentFragment( [] ), {
+					'marker-a': { start: [ 0 ], end: [ 0 ] }
+				} );
+
+				const expectedMarker = model.markers.get( 'marker-a' );
+
+				expect( getData( model ) ).to.equal( '<paragraph>fo[]o</paragraph>' );
+
+				expect( expectedMarker ).to.be.null;
+			} );
+
+			it( 'should not create marker when insertion is forbidden', () => {
+				// <wrapper><limit><paragraph>[]</paragraph></limit></wrapper>
+				// {<wrapper><limit><paragraph>foo</paragraph></limit></wrapper>}
```

Comment:
Here is a more interesting example but, in the end, similar to the one before. In this case it is more discussable whether the marker should be added or not.